### PR TITLE
ContentCAO: Fix broken attachments on join

### LINKED ISSUE #7841: Attachments not in effect on client if client joins after attachment is done
##### Issue type
 - Bug report
 
##### Minetest version
```
0.4.17.1
not in 5.0.0 (!)
```
Operating system: Ubuntu 16.04

##### Summary

I've written a mod(see "playertags").
It attaches nametag entities to people.
When opening a server, for me, the server admin, everything works perfect.
But not if I join from the same host : 
It seems that entity attachment is lost, the entity is just laying around, and not attached to the player when viewed from the client.

##### Steps to reproduce

Use playertags mod, host a game, join it from your local host, and see the problem.

### PR DESCRIPTION
**Pull request by SmallJoker**
**What happened:**
1) Object data is received. Client begins to read the data
2) Client initializes all its children (gob_cmd_update_infant)
3) Children try to attach to parent (yet not added)
4) Parent initializes, is added to the environment

And somewhere in between, Irrlicht wrecks up the attachments due to the missing matrix node

As a result, the attachments were just laying around at (0,0,0) without ever attaching to anything.

The solution here is to:
1) Use the same structure as `ServerActiveObject`
2) Attach all children _after_ the parent is _really_ initialized

## ToDo

**This PR is Ready for Review.**

If you have some time for testing: also check what happens when an entity dies.
Children attachments for players are kept, but lost for other entities.

## How to test
1) Install & enable [wield3d](https://forum.minetest.net/viewtopic.php?t=6407)
2) Start a server
3) Place a rail and put a cart on top
4) Attach to the cart
5) Join with a client. As soon they see the cart, the player must be inside
6) Punch the cart. If it drives away without the player this means you can trash this PR

Fixes #7841, #5075 (`GENERIC_CMD_PUNCHED` tweak, untested), partially #5013,

### CODE REVIEW

File: src/client/content_cao.cpp [hunk 1]
```diff
@@ -17,36 +17,35 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
-#include "serialization.h" // For decompressZlib
-#include "clientobject.h"
-#include "mesh.h"
-#include "itemdef.h"
-#include "tool.h"
 #include "content_cso.h"
-#include "sound.h"
-#include "nodedef.h"
+#include "environment.h"
+#include "itemdef.h"
```
Comment (SmallJoker) [inline src/client/content_cao.cpp]: Alphabetic sorting without functional changes.

File: src/client/content_cao.cpp [hunk 2]
```diff
@@ -443,47 +444,100 @@ scene::IAnimatedMeshSceneNode* GenericCAO::getAnimatedMeshSceneNode()
...
+		if (auto *o = m_env->getActiveObject(old_parent))
+			o->removeAttachmentChild(m_id);
+		if (parent)
+			parent->addAttachmentChild(m_id);
+	}
+
 	updateAttachments();
 }
 
+void GenericCAO::getAttachment(int *parent_id, std::string *bone, v3f *position,
```
Comment (SmallJoker) [inline src/client/content_cao.cpp]: Yet unused. Might be useful for CSMs later on.

File: src/client/content_cao.cpp [hunk 3]
```diff
@@ -723,7 +777,6 @@ void GenericCAO::addToScene(ITextureSource *tsrc)
 	updateNodePos();
 	updateAnimation();
 	updateBonePosition();
-	updateAttachments();
```
Comment (SmallJoker) [inline src/client/content_cao.cpp]: Moved to ClientEnvironment

File: src/client/clientenvironment.cpp [hunk 4]
```diff
@@ -392,7 +390,20 @@ void ClientEnvironment::addActiveObject(u16 id, u8 type,
 			<<std::endl;
 	}
 
-	addActiveObject(obj);
+	u16 new_id = addActiveObject(obj);
+
+	// Object initialized:
+	if ((obj = getActiveObject(new_id))) {
```
Comment (DS-Minetest) [inline src/client/clientenvironment.cpp]: Are the extra parenthesis needed?

Comment (SmallJoker) [reply src/client/clientenvironment.cpp]: Technically no, but there will be a compiler warning if left out.

File: src/client/content_cao.cpp [hunk 5]
```diff
@@ -443,47 +444,100 @@ scene::IAnimatedMeshSceneNode* GenericCAO::getAnimatedMeshSceneNode()
...
+}
+
+void GenericCAO::clearChildAttachments()
+{
+	// Cannot use for-loop here: setAttachment() modifies 'm_attachment_child_ids'!
+	while (!m_attachment_child_ids.empty()) {
+		int child_id = *m_attachment_child_ids.begin();
+
+		if (ClientActiveObject *child = m_env->getActiveObject(child_id))
+			child->setAttachment(0, "", v3f(0, 0, 0), v3f(0, 0, 0));
```
Comment (DS-Minetest) [inline src/client/content_cao.cpp]:
```suggestion
			child->setAttachment(0, "", v3f(0.0f, 0.0f, 0.0f), v3f(0.0f, 0.0f, 0.0f));
```
or
```suggestion
			child->setAttachment(0, "", v3f(), v3f());
```

File: src/client/content_cao.cpp [hunk 6]
```diff
@@ -443,47 +444,100 @@ scene::IAnimatedMeshSceneNode* GenericCAO::getAnimatedMeshSceneNode()
...
+		removeAttachmentChild(child_id);
+	}
+}
+
+void GenericCAO::clearParentAttachment()
+{
+	if (m_attachment_parent_id)
+		setAttachment(0, "", m_attachment_position, m_attachment_rotation);
+	else
+		setAttachment(0, "", v3f(0, 0, 0), v3f(0, 0, 0));
```
Comment (DS-Minetest) [inline src/client/content_cao.cpp]: Same as above.

File: src/client/content_cao.cpp [hunk 7]
```diff
@@ -443,47 +444,100 @@ scene::IAnimatedMeshSceneNode* GenericCAO::getAnimatedMeshSceneNode()
...
 
-	u16 attached_id = m_env->attachement_parent_ids[getId()];
+	if (m_attachment_parent_id)
+		obj = m_env->getActiveObject(m_attachment_parent_id);
 
-	if ((attached_id != 0) &&
-			(attached_id != getId())) {
-		obj = m_env->getActiveObject(attached_id);
-	}
 	return obj;
```
Comment (DS-Minetest) [inline src/client/content_cao.cpp]: Why not like this?
```cpp
	return m_attachment_parent_id ? m_env->getActiveObject(m_attachment_parent_id) : nullptr;
```

**Comment by LoneWolfHT** [comment]
When I was testing this in the CTF server I saw the hp bar flash in and out of existence (Above my head, where it should be) when I was changing my view mode with f7. Not sure if that helps...
EDIT:
image

**Comment by SmallJoker** [comment]
In some places the attachment adding is still out of order:
```
 setAttachment() id=2, parent=94557833757680, name=
   updateAttachments(): id=2, attach_id=1, obj=94557833757680
     matrix=0, node=94557830033744
Added 2 (2)
```

File: src/client/content_cao.cpp (hunk 7 above)
Comment (ClobberXD) [reply src/client/content_cao.cpp]: Just code style and readability, probably. Complex long conditions are less readable when using ternary operators.

Comment (ClobberXD) [reply src/client/content_cao.cpp]: If I'm not mistaken, the extra pair of parenthesis around the individual conditions aren't required.

**Comment by SmallJoker** [comment]
Mods used to test:

https://github.com/MT-CTF/capturetheflag/blob/master/mods/pvp/gauges/init.lua
 * Skip the range check at L30, `if false and ..`
https://github.com/Elkien3/playertag 
 * Comment out `minetest.after(10, step)` in init.lua L83


animation
^ Includes a little modification to make the gauge taller.

**Comment by SmallJoker** [comment]
@LoneWolfHT Fixed. Testing is welcome :D

**Comment by LoneWolfHT** [comment]
Will test tomorrow night if I get the time

**Comment by LoneWolfHT** [comment]
Tested on CTF while 2 other players were online. Seems to work. I'll test again when it gets more packed
Tested with carts and wield3d in singleplayer. Seems to work
:+1:

**Comment by ClobberXD** [comment]
Tested on CTF with around 12-14 players, all their attached entities showed up properly, and weren't spawned in the map block (chunk?) origin instead.

Also tested using carts and wield3d using a server+client and a client. The client has this bugfix applied, and the server host is attached to the cart. Punching the cart moves the player along with the cart, confirming that the player is indeed attached to the cart. This, obviously, isn't the case when testing with `master` - the cart moves separately, while the player is left hanging.

So this PR works, and does indeed fix the attachment bug. :+1:

**Comment by SmallJoker** [comment]
I appreciate your "real world" testing, it's very helpful to find out whether there are edge cases which weren't considered yet. Thanks to both of you.

Disclaimer: This does not fix the attachment bugs which arise by passing the 200m border. That's a different issue.

File: src/activeobject.h [hunk 8]
```diff
@@ -98,6 +100,16 @@ class ActiveObject
 
 
 	virtual bool collideWithObjects() const = 0;
+
+
+	virtual void setAttachment(int parent_id, const std::string &bone, v3f position,
+			v3f rotation) {}
+	virtual void getAttachment(int *parent_id, std::string *bone, v3f *position,
```
Comment (sfan5) [inline src/activeobject.h]: should be const method

File: src/client/clientobject.h [hunk 9]
```diff
@@ -51,8 +53,12 @@ class ClientActiveObject : public ActiveObject
 	virtual scene::ISceneNode *getSceneNode() { return NULL; }
 	virtual scene::IAnimatedMeshSceneNode *getAnimatedMeshSceneNode() { return NULL; }
 	virtual bool isLocalPlayer() const {return false;}
+
 	virtual ClientActiveObject *getParent() const { return nullptr; };
-	virtual void setAttachments() {}
+	virtual const std::unordered_set<int> &getAttachmentChildIds()
+	{ static std::unordered_set<int> rv; return rv; }
```
Comment (sfan5) [inline src/client/clientobject.h]: should be const method

File: src/client/content_cao.cpp [hunk 10]
```diff
@@ -1561,13 +1587,9 @@ void GenericCAO::processMessage(const std::string &data)
 		}
 	} else if (cmd == GENERIC_CMD_SPAWN_INFANT) {
 		u16 child_id = readU16(is);
-		u8 type = readU8(is);
+		//u8 type = readU8(is); maybe this will be useful later
```
Comment (sfan5) [inline src/client/content_cao.cpp]: since this <i>does</i> exist in the cmd, I'd prefer to read it and use `(void)` to silence the unused warning

**Pull request merged**